Attempt to parse a `Response[F]` when transmitting a request body fails in the Ember client

danicheg opened the pull request:
Resolves #7655

To illustrate the problem, let's glance at this sequence diagram:
```mermaid
sequenceDiagram
    participant client
    participant socket
    participant server
    client->>socket: writes a request body
    server->>socket: reads a request body
    server->>server: prepares a response
    server->>socket: writes a response
    client->>socket: reads a response
```
Here, `client` is an Ember client, `socket` is an `fs2.Socket`, and `server` is an arbitrary HTTP server. As per the current implementation (both in `series/0.23` and `1.0.x`), the client sequentially writes the request body to the Socket, and once done, reads a Response from the Socket.
The problem arises when the Server doesn't process the request body (fully or partially), responds eagerly, and closes the connection. In this situation, Ember's sequential processing fails with something like an `IOException: Broken pipe` exception on the JVM. According to the HTTP spec, Ember's sequential behaviour is truly legit (I guess many HTTP clients behave the same way). 
```mermaid
sequenceDiagram
    participant client
    participant socket
    participant server
    client->>socket: writes a request body
    server->>socket: reads a request body (up to the `max-content-length`, e.g.)
    server->>server: prepares a response
    server->>socket: eagerly writes a response
    client->>socket: still writes a request body
    Note right of client: here's where an `IOException` arises
    client->>socket: reads a response
    Note right of client: it doesn't happen actually
```
So, my suggestion in this PR is fairly straightforward — for specific errors coming from the Socket that can be identified as related to the closed connection, let's attempt to parse the Response anyway, to check whether it's present.
```mermaid
sequenceDiagram
    participant client
    participant socket
    participant server
    client->>socket: writes a request body
    server->>socket: reads a request body (up to the `max-content-length`, e.g.)
    server->>server: prepares a response
    server->>socket: eagerly writes a response
    client->>socket: still writes a request body
    Note right of client: now, we handle a raised `IOException`
    client->>socket: successfully reads a response
```
Surely, a smarter solution would be to implement something like simultaneous reading from the Socket while transmitting the request to it (which is also compliant with the HTTP spec). But that solution is **more invasive** and would likely require thorough testing, as it differs significantly from the current behaviour.

danicheg committed:
Add a test case for #7655 in the EmberServerSuite

danicheg committed:
Increase the size of payload

reardonj commented on ember-server/shared/src/test/scala/org/http4s/ember/server/EmberServerSuite.scala:
```diff
@@ -193,4 +196,19 @@ class EmberServerSuite extends Http4sSuite {
+    def runReq[A: EntityEncoder[IO, *]](server: Server, entity: A) = {
+      val req =
+        Request[IO](Method.POST, uri = url(server.addressIp4s, "/ignored-body/what"))
+          .withEntity(entity)
```
It looks like this test design is using a non-chunked body. My demo code had a lazy body, which caused the Ember client to send 2 packets. I think this is why you need a much larger body to reproduce.

danicheg committed:
Merge branch 'series/0.23' into test-case-for-#7655

danicheg committed:
Ignore 'java.io.IOException:Broken pipe' when writing to an fs2.Socket in the Ember client

The main reason for this is that Servers might respond eagerly without reading the entire request body. A 'java.io.IOException: Broken pipe' when writing to an fs2.Socket __likely__ means that a Server has already closed the connection.

danicheg committed:
Expand the number of ignored IOExceptions when writing to an fs2.Socket in the Ember client

danicheg committed:
Scalafix

danicheg committed:
Shorten the impl

danicheg committed:
Make the fix platform agnostic

danicheg committed:
Temporarily (or not) disable test for JS

danicheg committed:
Merge branch 'series/0.23' into test-case-for-#7655

danicheg commented on ember-server/shared/src/test/scala/org/http4s/ember/server/EmberServerSuite.scala:
```diff
@@ -193,4 +196,23 @@ class EmberServerSuite extends Http4sSuite {
+  fixture().test(
+    "#7655 — client shouldn't fail request processing when server ignores the request body"
+  ) { case (server, client) =>
+    assume(
+      Platform.isJvm || Platform.isNative,
+      "Currently, the mentioned behaviour is only supported on JVM/Native",
+    )
```
There's an ongoing discussion in discord about a discrepancy in the behaviour of `fs2.Socket` between JVM/Native and JS. There's a chance we've found an issue in fs2, and it might be resolved there, which would allow us to enable that test for the JS platform as well.

danicheg commented on ember-server/shared/src/test/scala/org/http4s/ember/server/EmberServerSuite.scala:
```diff
@@ -193,4 +196,23 @@ class EmberServerSuite extends Http4sSuite {
+      "Currently, the mentioned behaviour is only supported on JVM/Native",
+    )
+
+    val req =
+      Request[IO](
+        Method.POST,
+        uri = url(server.addressIp4s, "/ignored-body/hello"),
+        body = fs2.Stream.constant(42.toByte),
+      )
+
+    client
+      .expect[String](req)
+      .assertEquals("hello")
```
Without the fix to the Ember client, we’re getting the following error in this test:
```
==> X org.http4s.ember.server.EmberServerSuite.#7655 — client shouldn't fail request processing when server ignores the request body  0.014s java.io.IOException: Broken pipe
    at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
    at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:62)
    at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:132)
    at sun.nio.ch.IOUtil.write(IOUtil.java:97)
    at sun.nio.ch.IOUtil.write(IOUtil.java:53)
    at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:532)
    at fs2.io.net.SelectingSocket.$anonfun$write$1(SelectingSocket.scala:57)
    at delay @ fs2.io.net.SelectingSocket.go$1(SelectingSocket.scala:56)
    at flatMap @ fs2.io.net.SelectingSocket.go$1(SelectingSocket.scala:59)
    at delay @ fs2.io.net.SelectingSocket.write(SelectingSocket.scala:65)
    at flatMap @ fs2.io.net.SelectingSocket.write(SelectingSocket.scala:65)
    at getAndSet @ org.typelevel.keypool.KeyPool$.destroy(KeyPool.scala:119)
    at apply @ org.http4s.ember.server.EmberServerBuilder.$anonfun$build$2(EmberServerBuilder.scala:235)
    at use @ fs2.Compiler$Target.compile(Compiler.scala:158)
```

danicheg committed:
Use better naming in the ClientHelpersPlatform

ivan-klass commented:
Any updates on this? I would like to have it merged, what's the current blocker to accept these changes?
